torchaudio/functional.py

```diff
@@ -844,6 +844,147 @@ def compute_deltas(specgram, win_length=5, mode="replicate"):
 
 
 @torch.jit.script
+def gain(waveform, gain_db=1.0):
+    # type: (Tensor, float) -> Tensor
+    r"""Apply amplification or attenuation to the audio signal.
+
+    Args:
+       waveform (torch.Tensor): Tensor of audio of dimension (channel, time)
+       gain_db (float) Gain adjustment in decibels (dB) (Default: `1.0`).
+
+    Returns:
+       torch.Tensor: waveform amplified by gain_db
+    """
+    if (gain_db == 0):
+        return waveform
+
+    # gain_db = 20log(Tensor_output/tensor_input)
+    ratio = 10 ** (gain_db / 20)
+
+    return waveform * ratio
+
+
+@torch.jit.script
+def scale_to_interval(waveform, interval=1.0):
+    # type: (Tensor, float) -> Tensor
+    r"""Scales waveform to an interval.
+    Args:
+       waveform (torch.Tensor): Tensor of audio of dimension (channel, time)
+       interval (float): The bounds of the interval, where the float indicates
+           the upper bound and the negative of the float indicates the lower
+           bound (Default: `1.0`)
+           Example: interval=1.0 -> [-1.0, 1.0]
+    Returns:
+       torch.Tensor: waveform scaled to interval
+    """
+    abs_max = torch.max(torch.abs(waveform))
+    ratio = abs_max / interval
+    waveform /= ratio
+
+    assert torch.max(torch.abs(waveform)) == interval
+
+    return waveform
+
+
+def _add_noise_shaping(dithered_waveform, waveform):
+    r"""Noise shaping is calculated by error:
+    error[n] = dithered[n] - original[n]
+    noise_shaped_waveform[n] = dithered[n] + error[n-1]
+    """
+    error = dithered_waveform - waveform
+
+    # add error[n-1] to dithered_waveform[n], so offset the error by 1 index
+    for index in range(error.size()[0]):
+        err = error[index]
+        error_offset = torch.cat((torch.zeros(1), err))
+        error[index] = error_offset[:waveform.size()[1]]
+
+    noise_shaped = dithered_waveform + error
+    return noise_shaped
+
+
+def _get_rand_tensor(size):
+    # type: (int) -> int
+    r"""Returns tensor of a random int according to the size.
+    """
+    return torch.randint(size, [1, ]).item()
+
+
+@torch.jit.script
+def dither(waveform, probability_density_function="TPDF", noise_shaping=False, ns_filter=""):
+    # type: (Tensor, str, bool, str) -> Tensor
+    r"""Maximize the dynamic range of audio stored at a particular bit-depth.
```

Comment:
Could you expand a bit on this as to why adding that noise maximizes this?

Reply:
Ah, I got this from the [SoX documentation](http://sox.sourceforge.net/sox.html)

"Dithering is a technique used to maximise the dynamic range of audio stored at a particular bit-depth. Any distortion introduced by quantisation is decorrelated by adding a small amount of white noise to the signal. In most cases, SoX can determine whether the selected processing requires dither and will add it during output formatting if appropriate."

Reply:
It'd be interesting to just add a few lines explaining why this works